src/coreclr/jit/optimizer.cpp

```diff
@@ -5166,40 +5193,110 @@ bool Compiler::optIsLoopClonable(unsigned loopInd)
 #endif // JIT32_GCENCODER
     if (fgReturnCount + loopRetCount > epilogLimit)
     {
-        JITDUMP("Loop cloning: rejecting loop because it has %d returns; if added to previously-existing %d returns, "
-                "would exceed the limit of %d.\n",
-                loopRetCount, fgReturnCount, epilogLimit);
+        JITDUMP("Loop cloning: rejecting loop L%02u. It has %d returns;"
+                " if added to previously existing %d returns, it would exceed the limit of %d.\n",
+                loopInd, loopRetCount, fgReturnCount, epilogLimit);
+        return false;
+    }
+
+    unsigned ivLclNum = loop.lpIterVar();
+    if (lvaVarAddrExposed(ivLclNum))
+    {
+        JITDUMP("Loop cloning: rejecting loop L%02u. Rejected V%02u as iter var because is address-exposed.\n", loopInd,
+                ivLclNum);
+        return false;
+    }
+
+    BasicBlock* head = loop.lpHead;
+    BasicBlock* end  = loop.lpBottom;
+    BasicBlock* beg  = head->bbNext;
+
+    if (end->bbJumpKind != BBJ_COND)
+    {
+        JITDUMP("Loop cloning: rejecting loop L%02u. Couldn't find termination test.\n", loopInd);
+        return false;
+    }
+
+    if (end->bbJumpDest != beg)
+    {
+        JITDUMP("Loop cloning: rejecting loop L%02u. Branch at loop 'end' not looping to 'begin'.\n", loopInd);
+        return false;
+    }
+
+    // TODO-CQ: CLONE: Mark increasing or decreasing loops.
+    if ((loop.lpIterOper() != GT_ADD) || (loop.lpIterConst() != 1))
+    {
+        JITDUMP("Loop cloning: rejecting loop L%02u. Loop iteration operator not matching.\n", loopInd);
+        return false;
+    }
+
+    if ((loop.lpFlags & LPFLG_CONST_LIMIT) == 0 && (loop.lpFlags & LPFLG_VAR_LIMIT) == 0 &&
+        (loop.lpFlags & LPFLG_ARRLEN_LIMIT) == 0)
+    {
+        JITDUMP("Loop cloning: rejecting loop L%02u. Loop limit is neither constant, variable or array length.\n",
+                loopInd);
         return false;
     }
 
+    if (!(((loop.lpTestOper() == GT_LT || loop.lpTestOper() == GT_LE) && (loop.lpIterOper() == GT_ADD)) ||
```

Comment:
Seems like this is a prime candidate for `operIs(...)`

Reply:
I went ahead and created a new static OperIs() which can be used to (slightly) simplify this.